[LuoZYi] iP

LuoZYi opened the pull request:
### Duke ChatBot


Duke Chatbot is a convinient command-line task manager that empowers users to efficiently organize their tasks. With Duke, you can effortlessly add, list, mark, unmark, delete, and save tasks, all within the comfort of your command line. Duke also provides persistent task storage, ensuring that your tasks are preserved across sessions.

**Features**

- Intuitive Commands: Duke supports a variety of simple commands to manage your tasks.
- Task Listing: Quickly list all your tasks to get an overview.
- Task Marking: Mark tasks as done to keep track of your progress.
- Task Unmarking: Reverse the status of a previously marked task.
- Task Deletion: Easily remove tasks you no longer need.
- Task Searching: Find tasks that match specific keywords or descriptions.
- Task Types: Add different types of tasks, including to-do, event, and deadline tasks.
- Data Persistence: Duke saves your tasks to a file, ensuring you never lose your important data.
- User-Friendly Interface: Duke's user interface is designed for simplicity and ease of use.

**How to Get Started**
Clone or download the Duke Chatbot repository from [here](https://github.com/LuoZYi/ip).
Compile and run the Duke class.

**Usage**
Duke Chatbot understands the following commands:

1. bye: Exit the chatbot while saving your task list to a file.
2. list: List all tasks in your current task list.
3. mark <task_id>: Mark a task as done by specifying its ID.
4. unmark <task_id>: Unmark a previously marked task.
5. find <keyword>: Search for tasks that contain the specified keyword.
6. todo <description>: Add a to-do task with a description.
7. event <description> /from <datetime> /to <datetime>: Add an event task with a description, start date, and end date.
8. deadline <description> /by <datetime>: Add a deadline task with a description and due date.
9. delete <task_id>: Delete a task by its ID.

**Examples**
To add a to-do task: todo Buy groceries
To add an event task: event Birthday party /from 2023-10-01 18:00 /to 2023-10-01 22:00
To add a deadline task: deadline Submit report /by 2023-09-30 23:59
To mark a task as done: mark 1
To unmark a task: unmark 2
To delete a task: delete 3
To list all tasks: list


Feedback and Contributions
We welcome your feedback and contributions to improve Duke Chatbot. Feel free to [raise an issue](https://github.com/LuoZYi/ip/issues) or create a pull request.

Duke Chatbot is brought to you by the LuoZYi. Happy task managing! 🚀

"Duke Chatbot: Organize Your Life, One Task at a Time!"

Main Method
`public static void main(String[] args) {
        Duke myBot = new Duke();
        myBot.run();
    }`

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

LuoZYi committed:
Add Duke.java

LuoZYi committed:
Level 1

LuoZYi committed:
Level 2

LuoZYi committed:
Level 3

LuoZYi committed:
Level 3

LuoZYi committed:
Merge branch 'master' of https://github.com/LuoZYi/ip

LuoZYi committed:
text-ui-test

LuoZYi committed:
clear

LuoZYi committed:
a

LuoZYi committed:
Exception

LuoZYi committed:
Delete

LuoZYi committed:
Delete text-ui-test

LuoZYi committed:
Level-7

LuoZYi committed:
Delete some unnecessary lines

suryanshkushwaha commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+import javax.management.InstanceNotFoundException;
+import java.io.*;
```
I feel it might be better to import each required io libs separately.

suryanshkushwaha commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+        }
+        scanner.close();
+    }
+
+
+
+
```
Should there be one line gap between two functions?

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+public class Duke {
+    private String name = "Lakinta";
```
I think the name variable isn't changed throughout your program so it might be good if you declare this as a constant with the final keyword or if it's not used any other place aside from greeting function then can just put the name in the function :)

suryanshkushwaha commented on ToDo.java:
```diff
@@ -0,0 +1,14 @@
+
+    @Override
+    public String toString() {
+        String status = "[" + getStatusIcon() + "] ";
+        return String.valueOf(this.id) + ". " + "[T]" + status + this.description;
```
Could some lines would have been saved if we used 'super' keyword?

suryanshkushwaha commented on ToDo.java:
```diff
@@ -0,0 +1,14 @@
+        String status = "[" + getStatusIcon() + "] ";
+        return String.valueOf(this.id) + ". " + "[T]" + status + this.description;
+    }
+
+
```
Could these extra spaces be avoided?

suryanshkushwaha commented on Task.java:
```diff
@@ -0,0 +1,39 @@
+        System.out.println("Got it. I've added this task:");
+        System.out.println(this.toString().substring(2));
+    }
+}
```
Should a java file always end with a new blank line?

suryanshkushwaha reviewed (commented):
LGTM. Just some nits to fix here and there.
Good work done there1

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+                file.createNewFile();
+            }
+
+            try (FileOutputStream fos = new FileOutputStream(file);
+                 ObjectOutputStream oos = new ObjectOutputStream(fos)) {
+                // Write the taskArrayList to the file
+                oos.writeObject(taskArrayList);
+            } catch (IOException e) {
+                System.out.println("Error saving tasks to file!");
+                e.printStackTrace();
+            }
```
I think you can modify this inner try catch block to follow the coding standards as instructed on website!
![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/99804772/74483167-6524-4320-908b-3be5eccce5b1)

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    private void loadTasksFromFile() {
+        File file = new File(filePath);
+        if (file.exists()) {
+            try (FileInputStream fis = new FileInputStream(file);
+                 ObjectInputStream ois = new ObjectInputStream(fis)) {
+                // Read the taskArrayList from the file
+                taskArrayList = (ArrayList<Task>) ois.readObject();
+            } catch (IOException | ClassNotFoundException e) {
+                // Handle exceptions, e.g., if the file is corrupted
+                taskArrayList = new ArrayList<>();
+                e.printStackTrace();
+            }
+        } else {
+            taskArrayList = new ArrayList<>();
```
similar to the above comment for the try-catch block format and after thinking through, I think maybe you can consider moving taskArrayList = new ArrayList<>() before the if block just to make it cleaner! It's perfectly fine right now though.

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+                    }
+                }
+                else {
+                    //taskArrayList.add(new Task(response, count));
+
+                    CustomException exception = new CustomException();
+                    throw exception;
```
I think you can consider just throwing new CustomException() for cleaner code

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+
+    public void echo() {
+        while (true) {
+            String response = scanner.nextLine();
+            if (response.equals("bye")) {
+                exit();
+                break;
+            } else {
+                System.out.println(response);
+            }
+        }
+        scanner.close();
+    }
```
I think this is probably from the previous increments! Can remove if you're not using it anymore just for cleaning upp, similarly for chat() and addTask()

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+    private Scanner scanner = new Scanner(System.in);
+    private ArrayList<String> tasks = new ArrayList<>();
+    private ArrayList<Task> taskArrayList = new ArrayList<>();
+    private int count = 1;
```
I think since this count variable is now only called by the setTasks method, you can reduce the scope to follow coding standards (by initializing it inside the method itself) 
![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/99804772/ca15c966-c290-4cb8-9b7a-e8572fac7792)

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+
+    public Duke() {
+        // Load tasks from the file when the chatbot starts up
+        loadTasksFromFile();
```
ooh I like how you put this method inside of your constructor so that it's called first when Duke is initialized. I will take note of this for myself so thank you!

ph-nathan commented on Duke.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+
+    public void greeting() {
```
I think as for the naming convention, methods' names should be a verb so maybe you can consider greet() instead? (this is quite nitpicking though sorry haha)

ph-nathan commented on Event.java:
```diff
@@ -0,0 +1,19 @@
+    @Override
+    public String toString() {
+        String status = "[" + getStatusIcon() + "] ";
+        return String.valueOf(this.id) + ". " + "[E]" + status + this.description +
```
Nice job breaking into another line for improving code readability here. I think you can consider breaking before an operator just to follow the standard!
![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/99804772/7cd96cab-aa03-447a-a700-aa67f5dd9cbe)

ph-nathan reviewed (commented):
Overall, everything seems to be alright, just a few nits to fix and some cleaning up to do. Good job!

LuoZYi committed:
Level 8

LuoZYi committed:
A-MoreOOP

LuoZYi committed:
A-Package

LuoZYi committed:
A-Gradle

LuoZYi committed:
A-JUnit

LuoZYi committed:
gradle src java

LuoZYi committed:
Add Java docs

LuoZYi committed:
Follow coding standards

LuoZYi committed:
Level-9

LuoZYi committed:
Level-10

LuoZYi committed:
Add Assertions to Document Assumptions in Duke

Added assert statements in the Duke class to document and validate important assumptions at various points in the code.
The assertions cover conditions related to file paths, class initialization, method parameters, and exit procedures.
These assertions provide clarity about the expected behavior of the code and help ensure that assumptions hold true during program execution.
Assertions are a useful tool for enhancing code robustness and catching potential issues during development and testing.
Assertions are enabled during development and testing but should be disabled in production for improved performance.
This commit contributes to code quality and maintainability by explicitly documenting and verifying critical assumptions within the Duke class.

LuoZYi committed:
Merge pull request #2 from LuoZYi/branch-A-Assertions

Branch-A-Assertions

LuoZYi committed:
Improve Code Quality and Set Up Package GUI

In this commit, we have taken several steps to enhance the code quality and organization of the project:

1. Refactored and cleaned up code to improve readability and maintainability.
2. Reorganized classes and packages for better project structure.
3. Moved relevant files related to the graphical user interface (GUI) into a new 'GUI' package.
4. Updated import statements and references to reflect the new package structure.

These changes make the codebase more organized and lay the foundation for future enhancements and feature additions related to the GUI component of the application.

Changes Made:
- Refactored code for improved clarity and consistency.
- Created a new 'GUI' package and moved relevant files into it.
- Updated import statements and references accordingly.
- Reviewed and tested code changes to ensure they do not introduce errors.

LuoZYi committed:
Merge pull request #3 from LuoZYi/branch-A-CodingQuality

Improve Code Quality and Set Up Package GUI

LuoZYi committed:
Add Help extension

LuoZYi committed:
Merge pull request #4 from LuoZYi/branch-BCD-Extension

Add Help extension

LuoZYi committed:
A-UserGuide

LuoZYi committed:
Update MainWindow

LuoZYi committed:
Merge pull request #5 from LuoZYi/branch-BCD-Extension

Update MainWindow

LuoZYi committed:
Resolve conflict

LuoZYi committed:
Delete docs directory

LuoZYi committed:
Recover Accidentally deleted files

LuoZYi committed:
Resolve conflict

LuoZYi committed:
Resolve Conflict in location of files

LuoZYi committed:
Update README.md

LuoZYi committed:
Modify gradle location

LuoZYi committed:
Change Gradle location

LuoZYi committed:
Merge branch 'master' of https://github.com/LuoZYi/ip

LuoZYi committed:
remove unwanted readme and contributor

LuoZYi committed:
update gradle

LuoZYi committed:
modify build.gradle

LuoZYi committed:
Add submodule configuration for src/main/java

LuoZYi committed:
update gitmodule

LuoZYi committed:
update gitmodule

LuoZYi committed:
remove gitmodule

LuoZYi committed:
remove submodule src/main/java

LuoZYi committed:
recover java

LuoZYi committed:
modify userGuide

LuoZYi committed:
reduce long methods
